# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: Roger Hans Rodriguez Samanez   <br>
Url: https://github.com/.../03MAIR---Algoritmos-de-Optimizacion---2019/tree/master/SEMINARIO<br>
Problema:

(*) La respuesta es obligatoria
                                  

## Combinar cifras y operaciones

- El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva
- Disponemos de las 9 cifras del 1 al 9 (excluimos el 0) y de los 4 signos básicos de las operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/).
- Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad dada. Un ejemplo sería para obtener el 4:
$$4+2-6/3*1=4$$
- Debe analizarse el problema para encontrar todos los **valores enteros** posibles planteando las siguientes cuestiones:  
  A) ¿Qué valor **máximo** y **mínimo** se pueden obtener según las condiciones del problema?  
  B) ¿Es posible encontrar **todos los valores enteros posibles** entre dicho mínimo y máximo?

- Nota: Es posible usar la función de Python `eval` para evaluar una expresión:

```python
expresion = "4-2+6/3*1"
print(eval(expresion))  #4
```

## A) ¿Qué valor **máximo** y **mínimo** se pueden obtener según las condiciones del problema?  
Usaré fuerza bruta para poder explorar todo el espacio de soluciones. Este algoritmo nos permite para calcular el valor más alto y más bajo posible dadas las restricciones y poder responder a las preguntas planteadas:

In [10]:
from itertools import combinations, permutations
from sympy import sympify

# Parámetros base
digitos = '123456789'
operadores = ['+', '-', '*', '/']

# Resultados y contadores
resultados_enteros = {}          # Diccionario: resultado entero → expresiones que lo generan
expresiones_validas = []         # Lista de (expresión, resultado) válidos
total_evaluadas = 0              # Total de expresiones generadas

# Evaluación de todas las expresiones posibles
for comb in combinations(digitos, 5):
    for perm_digitos in permutations(comb):
        for perm_ops in permutations(operadores):
            expr = perm_digitos[0]
            for i in range(4):
                expr += f" {perm_ops[i]} {perm_digitos[i + 1]}"
            try:
                resultado = sympify(expr)
                total_evaluadas += 1
                if float(resultado).is_integer():
                    valor = int(resultado)
                    expresiones_validas.append((expr, valor))
                    if valor not in resultados_enteros:
                        resultados_enteros[valor] = []
                    resultados_enteros[valor].append(expr)
            except:
                continue

# Cálculo de mínimos, máximos y cobertura
valores_encontrados = set(resultados_enteros.keys())
minimo = min(valores_encontrados)
maximo = max(valores_encontrados)
rango_completo = set(range(minimo, maximo + 1))
faltantes = sorted(list(rango_completo - valores_encontrados))

# Resultados finales
print("Resultados del algoritmo")
print("---------------------------")
print(f"Total de expresiones evaluadas: {total_evaluadas}")
print(f"Total de expresiones con resultado entero: {len(expresiones_validas)}")
print(f"Total de valores enteros únicos obtenidos: {len(valores_encontrados)}")
print(f"Valor mínimo alcanzado: {minimo}")
print(f"Valor máximo alcanzado: {maximo}")
print(f"¿Rango completo entre mínimo y máximo?: {'Sí' if not faltantes else 'No'}")
if faltantes:
    print(f"Valores faltantes (primeros 10): {faltantes[:10]}")


Resultados del algoritmo
---------------------------
Total de expresiones evaluadas: 362880
Total de expresiones con resultado entero: 90000
Total de valores enteros únicos obtenidos: 147
Valor mínimo alcanzado: -69
Valor máximo alcanzado: 77
¿Rango completo entre mínimo y máximo?: Sí


Entonces se obtiene que:
 - Valor **mínimo** alcanzado: -69
 - Valor **máximo** alcanzado: 77

## B) ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo?
Del resultado en el punto anterior, se verifica que **SÍ** es posible encontrar todos los valores enteros entre el mínimo y el máximo

## 1. ¿Cuántas posibilidades hay *sin tener en cuenta las restricciones*?

Supongamos que **no aplicamos ninguna restricción**, es decir:

- Se permiten **repeticiones de dígitos** (puedo usar el mismo dígito más de una vez),
- Se pueden usar **cualquier combinación de operadores**, repitiéndolos si se desea,
- No se exige usar cada operador una sola vez.

Entonces:

- Hay 9 posibles dígitos (del 1 al 9),
- Usamos **5 posiciones de dígito**, con repetición permitida → \(9^5\) combinaciones de dígitos,
- Hay 4 posiciones para operadores (`op1`, `op2`, `op3`, `op4`),
- En cada una podemos poner cualquier operador entre `+ - * /` → \(4^4\) combinaciones de operadores.

#### Cálculo:

$$
\text{Total sin restricciones} = 9^5 \cdot 4^4 = 59,\!049 \cdot 256 = \boxed{15116544}
$$


---



## 2. ¿Cuántas posibilidades hay *teniendo en cuenta todas las restricciones*?

En este caso, **sí aplicamos todas las condiciones del enunciado**, que son:

- Se deben usar **5 dígitos únicos**, sin repetir, elegidos del conjunto `{1, 2, ..., 9}` → combinación sin repetición.
- Luego esos 5 dígitos se **permutan** en todas las posiciones posibles (porque el orden importa en la expresión).
- Debemos usar exactamente **los 4 operadores básicos**, **una sola vez cada uno**, en alguna de sus permutaciones posibles.
- Los resultados deben ser **números enteros**

### Detalle del cálculo:

1. **Elegir 5 dígitos distintos** de los 9 disponibles:
   - C(9, 5) = 126

2. **Permutar esos 5 dígitos**:
   - 5! = 120

3. **Permutar los 4 operadores diferentes** (uno de cada tipo):
   - 4! = 24

### Resultado:
- Total con restricciones **sin considerar números enteros** = C(9, 5) × 5! × 4! = 126 × 120 × 24 = **362880**
- Total con restricciones **considerando números enteros** = **90000**  (obtenido del punto A)


In [11]:
print(f"Total de expresiones con resultado entero: {len(expresiones_validas)}")

Total de expresiones con resultado entero: 90000


## Modelo para el espacio de soluciones.
## 3. ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)

#### Elección principal: **listas (`list`) y conjuntos (`set`)**

Durante el desarrollo del algoritmo, se han utilizado principalmente dos estructuras de datos fundamentales en Python:

1. **Listas (`list`)**:
   - Para representar las expresiones como secuencias ordenadas de elementos:  
     Ejemplo: `['4', '+', '2', '-', '6', '/', '3', '*', '1']`
   - Permiten construir dinámicamente cada expresión mediante `zip()`, concatenación y bucles.

2. **Conjuntos (`set`)**:
   - Para almacenar todos los **valores enteros únicos** obtenidos al evaluar las expresiones.
   - Se elige `set` porque:
     - Garantiza unicidad automáticamente (no permite duplicados),
     - Ofrece operaciones eficientes para verificar pertenencia (`in`) y calcular el rango de resultados.

Adicionalmente, para ciertos algoritmos de mejora se emplearon estructuras especializadas como colas de prioridad y se utilizaron diccionarios para organizar resultados, descartando otras opciones menos eficientes.

---

#### Estructura considerada inicialmente

Al inicio pensé en solo usar Listas como estructura de datos para almacenar tanto número como signos me dí cuenta de que era poco manejable al tratar las restricciones de números y signos a la vez.

---

#### Evolución hacia evaluación progresiva

Al identificar que las divisiones podían causar errores o floats no enteros, se optó por evaluar expresión por pasos, lo que requiere:

- Tomar los dígitos y operadores como elementos separados (listas),
- Aplicar operadores uno a uno (`a op b`) dentro de un bucle.

Esto **reforzó la elección de listas** como estructura principal para manipular la expresión antes de evaluarla.

---

#### Conclusión

| Necesidad                     | Estructura elegida |
|------------------------------|--------------------|
| Construcción de expresiones  | `list`             |
| Evaluación controlada        | `list`             |
| Almacenamiento de resultados | `set`              |
| Control de unicidad          | `set`              |

La elección de estas estructuras ha permitido que el algoritmo sea más claro, más controlable y más eficiente tanto en tiempo como en espacio, en comparación con trabajar directamente con strings o estructuras más complejas como árboles sintácticos (que no eran necesarios dada la ausencia de paréntesis en las operaciones).


## Según el modelo para el espacio de soluciones.
## 4. ¿Cual es la función objetivo? ¿Es un problema de maximización o minimización?
### Función objetivo
Este problema no se enmarca en un esquema tradicional de optimización con una función a minimizar o maximizar. Más bien, se trata de un problema de búsqueda combinatoria exhaustiva, donde el objetivo es generar todas las expresiones aritméticas válidas bajo ciertas restricciones y conservar aquellas cuyo resultado sea un número entero.

#### Formulación matemática de la función objetivo

Podemos representar la evaluación de cada expresión como una función:

> $f(a,b,c,d,e,\alpha, \beta, \gamma, \delta) = a\ \alpha\ b\ \beta\ c\ \gamma\ d\ \delta\ e$

**sujeta a:**

- $a, b, c, d, e \in \{1,2,3,4,5,6,7,8,9\}$, con $a \neq b \neq c \neq d \neq e$
- $\alpha, \beta, \gamma, \delta \in \{+, -, *, /\}$, con $\alpha \neq \beta \neq \gamma \neq \delta$
- Evaluación sin paréntesis, respetando la **precedencia estándar** de los operadores

Y además se impone:

> $f(a,b,c,d,e,\alpha, \beta, \gamma, \delta) \in \mathbb{Z}$

Esto implica que **solo se consideran expresiones cuyo resultado sea un número entero**, descartando aquellas que produzcan fracciones.

#### Enfoque práctico desde el curso

Tal como se plantea en las sesiones VC2–VC5 del curso:

> La función objetivo práctica es **evaluar todas las expresiones construidas bajo las restricciones** y conservar únicamente aquellas cuyo resultado pertenece a los enteros.

A partir de estos resultados, se puede:

- Determinar el **mínimo** y el **máximo** valor generado,
- Verificar si el conjunto de enteros resultantes **forma un rango continuo**.

Aunque no se trata de un problema típico de optimización matemática, sí se puede formalizar una función objetivo tanto **simbólica** como **algorítmica**, lo que permite analizar el espacio de soluciones, evaluar expresiones bajo restricciones estrictas, y aplicar conceptos clave del diseño de algoritmos.

### ¿Es un problema de maximización o minimización?
Este problema **no es un caso clásico de optimización**, ya que no se plantea de forma explícita una función a maximizar o minimizar bajo restricciones numéricas continuas.  
Aunque el objetivo no se define en términos de optimización clásica, puede reinterpretarse como una función simbólica cuyos resultados se filtran por pertenencia a ℤ (los enteros), sirviendo como métrica de evaluación de la expresión.

## 5. Diseña un algoritmo para resolver el problema por fuerza bruta

In [7]:
#Definición de cifras y operadores  
cifras = ['1','2','3','4','5','6','7','8','9']
operadores = ['+','-','/','*']

#Cálculo de las variaciones sin repetición de una lista
def variaciones_sin_repeticion(lista, k):

  if k == 0:
    return [[]]
  variaciones_resultantes = []
  for i in range(len(lista)):
    elemento = lista[i]
    sublista = lista[:i] + lista[i+1:]
    variaciones = variaciones_sin_repeticion(sublista,k-1)
    for v in variaciones:
      variaciones_resultantes.append([elemento]+v)
  return variaciones_resultantes
#Cálculo de las combinaciones entre cifras y operadores  
#Las cifras ocupan posciones impares y los operadores posiciones pares
def combinar_numeros_signos(cifras,operadores):
  expresion = ''
  for i in range(len(operadores)):
    expresion += cifras[i]
    expresion += operadores [i]
  expresion += cifras[-1]
  return expresion

#Para cierto valor, buscamos una expresión
def buscar_expresion_fuerza_bruta(cifras, operadores, valor):
  #Calculamos las variaciones sin repetición de los números y signos
  var_numeros = variaciones_sin_repeticion(cifras,5)
  var_signos = variaciones_sin_repeticion(operadores,4)

  #Por cada variación realizamos una combinación de números y signos y comprobamos el resultado
  for n in var_numeros:
    for s in var_signos:
      expresion = combinar_numeros_signos(n,s)
      if eval(expresion) == valor:
        return expresion
  return ''

In [8]:
valor = 4
print("La expresión asociada al valor", valor, "es: ", buscar_expresion_fuerza_bruta(cifras,operadores,valor))

La expresión asociada al valor 4 es:  1-2*3/6+4


## 6. Calcula la complejidad del algoritmo por fuerza bruta

El número total de expresiones que el algoritmo evalúa viene dado por:

1. **Combinaciones de 5 dígitos únicos** tomados del conjunto {1, ..., 9}:  
   C(9, 5) = 126

2. **Permutaciones de esos 5 dígitos** (el orden importa en la expresión):  
   5! = 120

3. **Permutaciones de los 4 operadores distintos** (+, -, *, /):  
   4! = 24


### Complejidad asintótica:

O(C(9, 5) × 5! × 4!) = **O(362,880)**  =>  **$O(n!)$.**

Este valor es constante porque el conjunto de dígitos y operadores está fijo, pero el número de combinaciones sigue siendo considerable. Aun así, puede evaluarse en tiempo aceptable con un algoritmo eficiente.
Se puede concluir que la complejidad total del algoritmo es de orden factorial **$O(n!)$.**




## 7. Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

### BACKTRACKING

En este caso, para poder mejorar la eficiencia del algoritmo de fuerza bruta implementado anteriormente, usaremos un algoritmo de backtracking donde iremos ramificando cada uno de los números posibles para cada uno de los operadores. Podemos ver un ejemplo del árbol siguiente: 




                                       1
                          ┌────────────┼────────────┬────────────┐
                         [+]          [-]          [*]          [/]
                         / \           ...           ...           ...
                       2     6
                     /   \
                   [-]   [*] ...
                  /   \
                 3    ...
                  \
                   [*]  
                 /  |  \
                4   5   6
               /    |    \
             [/]    ...    ...
            / | \  
           5  6  7  ...
              

En esta figura podemos ver como se desarrollaría el árbol para cada una de las posibilidades usando el ejemplo propuesto anteriormente de:

```python
cifras = [1, 2, 3, 4, 5, 6, 7, 8, 9]
operadores = ['+', '-', '*', '/']
```

Para el desarrollo del algoritmo, primero iteramos sobre la lista de números (ramificamos los números), comprobamos si la expresión que ya tenemos es válida, en caso contrario añadimos un operador y volvemos a iterar sobre los números restantes haciendo una llamada recursiva a la función.

En el caso de que se haya llegado al final del árbol, no se pueden utilizar más operadores se vuelve hacia atrás para probar con otra solución. Por lo tanto, esta es una estructura de backtracking.

In [45]:
def busqueda_expresion_backtracking(expresion, numeros, operadores, valor):
    # Ramificar para añadir una cifra a la expresión
    for numero in numeros:
        numeros_restantes = numeros.copy()
        numeros_restantes.remove(numero)
        expresion_numero = f'{expresion}{numero}'

        # Comprobar si la expresión actual es una solución válida.
        if len(expresion_numero)>0 and eval(expresion_numero) == valor and len(operadores) == 0:
            return expresion_numero # Solución encontrada!

        # Retornamos 0 si ya hemos usado todos los operadores y hemos probado todos los números posibles
        elif len(operadores) == 0 and numero == numeros[-1]:
            return 0

        # Ramificamos para añadir un operador a la expresión
        for operador in operadores:
            nueva_expresion = f'{expresion_numero}{operador}'
            operadores_restantes = operadores.copy()
            operadores_restantes.remove(operador)

            mejor_solucion = busqueda_expresion_backtracking(nueva_expresion, numeros_restantes, operadores_restantes, valor)

            # En caso de tener una solución, el valor retornado será la cadena con la operación matemática
            if mejor_solucion != 0:
                return mejor_solucion
    return 0

In [74]:
#Ejemplo de backtracking
soluciones = busqueda_expresion_backtracking('',cifras,operadores,44)
print(soluciones)

1+5*9-4/2


### RAMIFICACIÓN Y PODA

En esta implementación se utilizan ciertos patrones previamente analizados con el objetivo de obtener los valores máximo y mínimo sin necesidad de recorrer exhaustivamente todo el espacio de soluciones posibles. Para ello, se aplican reglas que permiten **descartar ramas del espacio de búsqueda** que no contribuyen a alcanzar dichos extremos.

Cabe resaltar que esta optimización ha sido diseñada de forma específica para el conjunto de números y operadores considerados en este caso particular. Aún no se ha validado si esta estrategia se puede generalizar a otros conjuntos de entrada; por tanto, **no se trata de una mejora universal o generalista**. Este aspecto queda como línea futura de estudio.

Por otro lado, si se profundizara el análisis de las características de las expresiones que suelen producir los valores extremos (máximos y mínimos), sería posible **reducir aún más el espacio de búsqueda**, aplicando reglas más precisas o específicas de poda.

In [95]:
import itertools

# Evaluar y seleccionar un subconjunto de números más prometedores
def evaluar_numeros_candidatos(nums=('1','2','3','4','5','6','7','8','9')):
    nums = sorted([int(n) for n in nums])
    numeros_pares = [n for n in nums if n % 2 == 0]

    candidatos = []
    candidatos.append(nums[0])               # Agrega el menor número
    candidatos.extend(nums[-3:])             # Agrega los 3 mayores
    candidatos.extend(numeros_pares[:2])     # Agrega los 2 primeros pares (si existen)
    
    return sorted(set(candidatos))           # Elimina duplicados y ordena

# Generar todas las expresiones válidas con los candidatos y operaciones
def secuencias_operacionales_con_poda(nums=('1','2','3','4','5','6','7','8','9'), 
                                      operations=('+', '-', '/', '*')):
    # Obtener candidatos como strings
    nums_filtrados = tuple(str(n) for n in evaluar_numeros_candidatos(nums))

    # Permutaciones de 5 cifras y de los 4 operadores
    permutaciones_numeros = itertools.permutations(nums_filtrados, 5)
    permutaciones_operadores = itertools.permutations(operations, 4)

    expresiones_validas = []

    for operadores in permutaciones_operadores:
        for numeros in permutaciones_numeros:
            expresion = []

            # Construir la expresión alternando número-operador
            for i in range(4):
                expresion.append(numeros[i])
                expresion.append(operadores[i])
            expresion.append(numeros[-1])  # Último número

            expresion_str = ''.join(expresion)

            try:
                resultado = eval(expresion_str)
                if float(resultado).is_integer():
                    expresiones_validas.append(expresion_str)
            except:
                continue  # Saltar divisiones por cero u otros errores

    return expresiones_validas

# Ejecución del algoritmo
expresiones_1 = secuencias_operacionales_con_poda()
print(len(expresiones_1))


264


In [107]:
def buscar_expresion_por_valor_ramificacion(valor_objetivo, 
                                            nums=('1','2','3','4','5','6','7','8','9'), 
                                            operations=('+', '-', '/', '*')):

    import itertools

    nums = tuple([str(n) for n in evaluar_numeros_candidatos(nums)])

    permuted_nums = tuple(itertools.permutations(nums, 5))
    permuted_operations = tuple(itertools.permutations(operations, len(operations)))

    for operation_sec in permuted_operations:
        for num_sec in permuted_nums:

            expresion = [None] * (len(num_sec) + len(operation_sec))
            i = j = k = 0

            while k < len(expresion):
                if k % 2 == 0:
                    expresion[k] = num_sec[i]
                    i += 1
                else:
                    expresion[k] = operation_sec[j]
                    j += 1
                k += 1

            expresion_str = ''.join(expresion)

            try:
                resultado = eval(expresion_str)
                if float(resultado).is_integer() and int(resultado) == valor_objetivo:
                    return expresion_str
            except ZeroDivisionError:
                continue
            except:
                continue

    return ''  # Si no se encontró ninguna


In [108]:
expresion = buscar_expresion_por_valor_ramificacion(4)
print(expresion)

1+7-2/4*8


La técnica de ramificación y poda permite reducir significativamente el espacio de búsqueda al aplicar criterios previos que descartan combinaciones poco prometedoras. En este caso, se seleccionan únicamente ciertos dígitos estratégicos y se generan expresiones válidas con operadores sin repetir. Aunque esta solución no garantiza explorar todo el espacio, es eficaz para encontrar resultados extremos (máximos o mínimos) con menor costo computacional. Es una estrategia eficiente, pero especializada y dependiente del conjunto de entrada.

### HEURÍSTICA

Para resolver este problema, se plantea una heurística que mejora el enfoque de fuerza bruta, reduciendo significativamente el tiempo de búsqueda. El algoritmo emplea una cola con prioridad, donde cada nodo representa una expresión matemática parcial construida a partir de dígitos y operadores no repetidos.

El proceso inicia con un nodo raíz vacío, que se encola con prioridad cero. Mientras la cola no esté vacía, se extrae el nodo con menor distancia al valor objetivo. Luego, se generan sus hijos combinando los dígitos y operadores aún disponibles. Si la longitud de la expresión es par (termina en un operador), se evalúa la parte numérica sin incluir ese último operador. Si es impar (expresión completa), se evalúa directamente.

En función del resultado, si es un número entero, el hijo se encola con una prioridad basada en su distancia al objetivo. Si el resultado es decimal, se le asigna una prioridad infinita, disminuyendo así sus posibilidades de ser explorado antes.

En conjunto, esta heurística mejora el rendimiento general priorizando caminos prometedores y evaluando expresiones parciales de forma anticipada, lo que permite reducir el espacio de búsqueda sin perder exactitud.

In [49]:
#import de librería de colas como estructura de datos
import queue 

In [50]:
# Genera los hijos de un nodo actual según su longitud
def crear_hijos(nodo, numeros, signos):
    hijos = []  # Lista para almacenar los hijos generados

    if len(nodo) % 2 == 0:
        # Si la longitud del nodo es par, se debe agregar un número no repetido
        for n in numeros:
            if n not in nodo:
                hijos.append(nodo + n)
    else:
        # Si la longitud es impar, se debe agregar un operador no repetido
        for s in signos:
            if s not in nodo:
                hijos.append(nodo + s)

    return hijos


In [58]:
# Algoritmo heurístico que mejora la búsqueda por fuerza bruta
def buscar_expresion_heuristico(numeros, signos, valor_objetivo):
    nodo_raiz = ''  # Nodo inicial (expresión vacía)
    nodos = queue.PriorityQueue()  # Cola con prioridad para ordenar los nodos
    nodos.put((0, nodo_raiz))  # Inicialización del nodo raíz con prioridad 0

    while not nodos.empty():
        nodo_actual = nodos.get()  # Seleccionamos el nodo de mayor prioridad (más prometedor)
        hijos = crear_hijos(nodo_actual[1], numeros, signos)  # Generamos los hijos del nodo actual

        for h in hijos:
            # Evaluamos la expresión según la longitud (evitamos terminar en un operador)
            if len(h) % 2 == 0:
                valor_hijo = eval(h[:len(h)-1])  # Si termina en operador, evaluamos sin el último carácter
            else:
                valor_hijo = eval(h)  # Si termina en número, evaluamos la expresión completa

            # Si el valor es entero, lo agregamos con su distancia al objetivo como prioridad
            if not isinstance(valor_hijo, float):
                nodos.put((abs(valor_objetivo - valor_hijo), h))
            else:
                # Si es decimal, se agrega con prioridad infinita (baja prioridad)
                nodos.put((float('inf'), h))

            # Si la expresión está completa y coincide con el valor deseado, se retorna como solución
            if len(h) == len(numeros) and valor_hijo == valor_objetivo:
                return h

    return ''  # Si no se encuentra solución

In [59]:
numeros = ['1','2','3','4','5','6','7','8','9']
signos = ['+','-','/','*']

In [60]:
#Heurística total:
for i in range(int(minimo), int(maximo)+1):
  print("La expresión asociada al valor", i, "es: ", buscar_expresion_heuristico(numeros,signos,i))

La expresión asociada al valor -69 es:  1-9*8+4/2
La expresión asociada al valor -68 es:  1-9*8+6/2
La expresión asociada al valor -67 es:  2-9*8+3/1
La expresión asociada al valor -66 es:  2-9*8+4/1
La expresión asociada al valor -65 es:  2-9*8+5/1
La expresión asociada al valor -64 es:  2-9*8+6/1
La expresión asociada al valor -63 es:  2-9*8+7/1
La expresión asociada al valor -62 es:  3-9*8+7/1
La expresión asociada al valor -61 es:  4-9*8+7/1
La expresión asociada al valor -60 es:  1-9*7+4/2
La expresión asociada al valor -59 es:  1-9*7+6/2
La expresión asociada al valor -58 es:  1-9*7+8/2
La expresión asociada al valor -57 es:  2-9*7+4/1
La expresión asociada al valor -56 es:  2-9*7+5/1
La expresión asociada al valor -55 es:  2-9*7+6/1
La expresión asociada al valor -54 es:  3-9*7+6/1
La expresión asociada al valor -53 es:  1-8*7+4/2
La expresión asociada al valor -52 es:  1-8*7+6/2
La expresión asociada al valor -51 es:  1-9*6+4/2
La expresión asociada al valor -50 es:  2-9*6+8/4


La heurística implementada representa una mejora significativa respecto al enfoque por fuerza bruta, ya que reduce la cantidad de nodos que es necesario explorar para encontrar una solución válida. Mientras que el algoritmo de fuerza bruta examina exhaustivamente todas las combinaciones posibles de cifras y operadores —lo que puede resultar costoso en términos de tiempo—, la heurística prioriza los nodos más prometedores y retrasa o descarta aquellos que tienen menor probabilidad de conducir a una solución.

En particular, los nodos que generan resultados poco útiles (como números decimales) son enviados a la cola con prioridad infinita, lo que disminuye su importancia en el orden de evaluación. Esto permite concentrar los recursos computacionales en rutas más efectivas, acelerando el proceso de búsqueda.

Asimismo, la evaluación anticipada de las expresiones permite identificar y filtrar de forma temprana aquellas ramas que no producirán resultados enteros, evitando así expandir nodos irrelevantes. En conjunto, esta estrategia mejora el rendimiento general del algoritmo, disminuyendo el tiempo requerido para encontrar una solución sin sacrificar exhaustividad.

In [98]:
#importamos la librería time para poder señalar marcas de tiempo
import time

In [99]:
#establecemos un valor cualquiera
valor = 44

In [100]:
#calculamos lo que toma a Fuerza Bruta calcular la expresión correcta
start = time.time()
expresion = buscar_expresion_FB(cifras, operadores, valor)
end = time.time()
metodo = "Fuerza Bruta"
print(f"[{metodo}] f(x) = {valor} → {expresion} | Tiempo: {end - start} segundos")


[Fuerza Bruta] f(x) = 44 → 1-4/2+5*9 | Tiempo: 0.24588608741760254 segundos


In [112]:
#calculamos lo que toma a Ramificación y Poda calcular la expresión correcta
start = time.time()
expresion = buscar_expresion_por_valor_ramificacion(valor,numeros,signos)
end = time.time()
metodo = "Ramificación y Poda"    
print(f"[{metodo}] f(x) = {valor} → {expresion} | Tiempo: {end - start} segundos")


[Ramificación y Poda] f(x) = 44 →  | Tiempo: 0.2779998779296875 segundos


In [113]:
#calculamos lo que toma a Backtracking calcular la expresión correcta
start = time.time()
expresion= busqueda_expresion_backtracking('',cifras, operadores, valor)
end = time.time()
metodo = "Backtracking"
print(f"[{metodo}] f(x) = {valor} → {expresion} | Tiempo: {end - start} segundos")


[Backtracking] f(x) = 44 → 1+5*9-4/2 | Tiempo: 0.0743098258972168 segundos


In [114]:
#calculamos lo que toma a Heuritico calcular la expresión correcta
start = time.time()
expresion = buscar_expresion_heuristico(numeros,signos,valor)
end = time.time()
metodo = "Heurística"
print(f"[{metodo}] f(x) = {valor} → {expresion} | Tiempo: {end - start} segundos")

[Heurística] f(x) = 44 → 9*5+1-4/2 | Tiempo: 0.003032207489013672 segundos


## 8. Calcula la complejidad del algoritmo

 

### RAMIFICACIÓN Y PODA

Gracias a esta estrategia de ramificación y poda basada en patrones, el algoritmo no escala en complejidad con el número total de cifras disponibles, ya que selecciona una cantidad fija de dígitos —denotada como `a`— para construir las expresiones. Por tanto, el número de combinaciones de cifras es constante en esta implementación.

Sin embargo, la complejidad sigue estando influenciada por la cantidad de operadores aritméticos disponibles (`n`). Dado que se generan todas las permutaciones posibles de estos operadores, el crecimiento sigue siendo factorial en `n`**, aunque en la práctica solo se usan los operadores `+`, `-`, `*` y `/`.

Asumiendo que:
- `a` = cantidad fija de dígitos seleccionados (constante),
- `n` = número de operadores distintos disponibles (variable),

Entonces, la complejidad del algoritmo se puede expresar como: **$O(a \cdot n!)$** en función del número de operadores.


### BACKTRACKING

La complejidad de este algoritmo en este caso va a depender obviamente del tamaño del espacio de búsqueda. En este caso tenemos el conjunto de números y operadores.

Suponiendo que tenemos N números y M operadores, en el peor de los casos, el algoritmo debe explorar todo el espacio de búsqueda, lo que implica visitar cada nodo del árbol de backtracking. Por lo tanto, la complejidad del algoritmo de backtracking es exponencial en el peor de los casos, $O(N^M)$ o $O(2^n)$.

### HEURÍSTICA

Al tratarse de una heurística, la complejidad no se puede calcular de forma directa. Las heurísticas son métodos aproximados que se utilizan para resolver problemas cuando no es posible encontrar una solución exacta en tiempo razonable. Las heurísticas se basan en la intuición y no se pueden analizar de la misma manera que los algoritmos exactos. En lugar de hablar de complejidad, es más común hablar de la calidad en relación al tiempo y espacio utilizado. En el punto anterior se puede ver como la heurística mejora los tiempos con respecto a fuerza bruta.

## 9. Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios. Aplica el algoritmo al juego de datos generado

In [77]:
#import para aleatoreidad
import random

In [78]:
#Generamos unos datos de entrada aleatorios: Números enteros entre el máximo y el mínimo
datos_aleatorios = []
num_datos = 5

for i in range(num_datos):
  datos_aleatorios.append(random.randint(minimo,maximo))

datos_aleatorios

[29, 72, 63, -68, -16]

In [115]:
for dato in datos_aleatorios:
  print("--------------------------------------------------")

  start = time.time()
  expresion = buscar_expresion_FB(numeros,signos,dato)
  end = time.time()
  metodo = "Fuerza Bruta"
  print(f"[{metodo}] f(x) = {valor} → {expresion} | Tiempo: {end - start} segundos")

  start = time.time()
  expresion = buscar_expresion_por_valor_ramificacion(dato, numeros,signos)
  end = time.time()
  metodo = "Ramificación y Poda"    
  print(f"[{metodo}] f(x) = {valor} → {expresion} | Tiempo: {end - start} segundos")
    
  start = time.time()
  expresion = busqueda_expresion_backtracking('', numeros,signos,dato)
  end = time.time()
  metodo = "Backtracking"    
  print(f"[{metodo}] f(x) = {valor} → {expresion} | Tiempo: {end - start} segundos")

  start = time.time()
  expresion = buscar_expresion_heuristico(numeros,signos,dato)
  end = time.time()
  metodo = "Heurística"  
  print(f"[{metodo}]   f(x) = {valor} → {expresion} | Tiempo: {end - start} segundos")



--------------------------------------------------
[Fuerza Bruta] f(x) = 44 → 1-4/2+5*6 | Tiempo: 0.2510676383972168 segundos
[Ramificación y Poda] f(x) = 44 → 9+4/1*7-8 | Tiempo: 0.06555318832397461 segundos
[Backtracking] f(x) = 44 → 1+5*6-4/2 | Tiempo: 0.06959843635559082 segundos
[Heurística]   f(x) = 44 → 9*3+4-2/1 | Tiempo: 0.0014529228210449219 segundos
--------------------------------------------------
[Fuerza Bruta] f(x) = 44 → 2-6/3+8*9 | Tiempo: 0.7572181224822998 segundos
[Ramificación y Poda] f(x) = 44 →  | Tiempo: 0.22646331787109375 segundos
[Backtracking] f(x) = 44 → 2+8*9-6/3 | Tiempo: 0.6793034076690674 segundos
[Heurística]   f(x) = 44 → 9*8+3-6/2 | Tiempo: 0.002714395523071289 segundos
--------------------------------------------------
[Fuerza Bruta] f(x) = 44 → 2-6/3+7*9 | Tiempo: 0.7489900588989258 segundos
[Ramificación y Poda] f(x) = 44 → 9+7/1*8-2 | Tiempo: 0.0344085693359375 segundos
[Backtracking] f(x) = 44 → 2+7*9-6/3 | Tiempo: 0.661017894744873 segundos
[He

## 10. Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

## Referencias

- Cormen, T. H., Leiserson, C. E., Rivest, R. L., & Stein, C. (2009). *Introduction to algorithms* (3rd ed.). MIT Press.

- Skiena, S. S. (2008). *The algorithm design manual* (2nd ed.). Springer.

- SymPy Development Team. (n.d.). *SymPy: Python library for symbolic mathematics*. https://docs.sympy.org/latest/index.html

- Python Software Foundation. (n.d.). *Operator precedence*. In *Python 3 documentation*. https://docs.python.org/3/reference/expressions.html#operator-precedence

- Diapositivas de clase


## 11. Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

## Líneas de avance en el estudio del problema

A medida que aumenta el tamaño del espacio de búsqueda —ya sea por la inclusión de más dígitos, más operadores o reglas más flexibles— el enfoque por fuerza bruta se vuelve rápidamente ineficiente. Por ello, es necesario explorar técnicas que permitan reducir el costo computacional mediante una búsqueda más inteligente y estructurada.

Durante el desarrollo de este trabajo, se han aplicado cuatro enfoques principales: **fuerza bruta**, **ramificación y poda** **backtracking** y una **heurística basada en prioridad**. Cada uno permitió obtener mejores resultados en términos de eficiencia frente al crecimiento del problema.

### Posibles caminos de mejora

1. **Optimización del espacio de búsqueda**  
   Mediante poda temprana y ordenación inteligente de los nodos, es posible evitar evaluar expresiones que, por su estructura, no podrán generar una solución válida. Este principio se puede extender a través de algoritmos como ramificación y poda, cuidando especialmente los casos donde el orden de operadores puede afectar significativamente el resultado.

2. **Variaciones del problema**  
   El problema puede ampliarse o modificarse para aumentar su complejidad:
   - Permitir la repetición de dígitos u operadores.
   - Aumentar el conjunto de operadores (por ejemplo, incluir potencias o módulo).
   - Permitir el uso de paréntesis o la concatenación de dígitos para formar números más grandes.

3. **Variación del objetivo**  
   En lugar de buscar una única expresión que genere un número específico, se puede plantear:
   - Encontrar todas las expresiones posibles que generen dicho número.
   - Determinar si se puede cubrir un rango completo de enteros (como se resolvió con backtracking).
   - Buscar expresiones que maximicen o minimicen el valor bajo restricciones más generales.

4. **Aplicación de heurísticas más sofisticadas**  
   Técnicas heurísticas permiten reducir el espacio de búsqueda priorizando ramas prometedoras. Este enfoque puede mejorarse con:
   - Funciones de evaluación más informadas.
   - Métodos como best-first search, beam search, o incluso A\* si se define un buen heurístico admisible.

5. **Implementación de algoritmos avanzados**  
   Además de lo revisado, se podrían explorar:
   - Algoritmos genéticos, que pueden explorar espacios grandes generando soluciones aproximadas.
   - Métodos de aprendizaje automático, para predecir qué combinaciones tienen mayor probabilidad de éxito.

### Conclusión

Avanzar en el estudio de este problema requiere un análisis profundo del espacio de soluciones y de las restricciones impuestas. A través de la incorporación de nuevas reglas, objetivos alternativos y técnicas más avanzadas, es posible ampliar significativamente el alcance del problema, así como mejorar la eficiencia de su resolución.
